compiler/rustc_resolve/src/imports.rs

```diff
@@ -926,11 +934,34 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                     let msg = "glob import doesn't reexport anything because no candidate is public enough";
                     self.lint_buffer.buffer_lint(UNUSED_IMPORTS, id, import.span, msg);
                 }
+                    self.privacy_errors.append(&mut privacy_errors);
                     return None;
                 }
                 _ => unreachable!(),
             };
 
+        let mut res = None;
+        if !self.privacy_errors.is_empty() {
+            // Get the Res for the last element, so that we can point to alternative ways of
+            // importing it if available.
+            let mut path = import.module_path.clone();
+            path.push(Segment::from_ident(ident));
+            if let PathResult::Module(ModuleOrUniformRoot::Module(module)) = self.resolve_path(
+                &path,
+                None,
+                &import.parent_scope,
+                Some(Finalize { report_private: ReportPrivate::Skip, ..finalize }),
```

Comment:
Not clear why introducing `ReportPrivate::Skip` is necessary.
Aren't these errors already deduplicated due to `dedup_span`?

Reply:
Calling `resolve_path` with `ReportPrivate::Error` will bail early on resolution errors and not return the `Res`/`DefId` we need. Calling it with `ReportPrivate::Report` will give us back the `Res` but will also produce spurious errors.

Reply:
This is speculative resolution, just to get the resolution result, right?
In that case it doesn't need `Finalize` at all, and `None` can be used here.

Reply:
Using `None` [worsens the case for sealed traits in foreign crates](https://github.com/rust-lang/rust/pull/112686/commits/b1fd6ac37810a95a4de6af63cbbb4f09f4d355c0#diff-3d63921e1b98fed9fd7bc43bdf56adb40750b22cfbbb9b7919c5a7b5d5197e82). If you prefer we can land the other changes, but reduces the benefit of that part of this change significantly, since the people who would need that note would be when encountering it in dependencies only. Maybe there's an alternative targeted tweak in `resolve_path` that we can keep the desired behavior for foreign crates too?

Reply:
Looks like `None` always behaves like `ReportPrivate::Error` here which is not what we need.
(I don't think it correlates with items being from foreign crates though.)

Reply:
I'll think what to do with this.

Reply:
Yeah, you're right it's not because of foreign crates, it's because of the behavior around imports. The incorrect `if !self.privacy_errors.is_empty()` was making this new logic not trigger anymore. It now points at the module immediately after, but that's something I can look into after rebasing on top of your latest changes.